core/grants.go

```diff
@@ -0,0 +1,187 @@
+
+	"github.com/golang/protobuf/proto"
+
+	"chain/errors"
+	"chain/net/http/authz"
+	"chain/net/http/httpjson"
+)
+
+// an api-friendly representation of a grant
+type apiGrant struct {
+	GuardType string                 `json:"guard_type"`
+	GuardData map[string]interface{} `json:"guard_data"`
+	Policy    string                 `json:"policy"`
+	CreatedAt string                 `json:"created_at"`
+}
+
+// ErrMissingTokenID is returned when a token does not exist.
+var errMissingTokenID = errors.New("id does not exist")
+
+func (a *API) createGrant(ctx context.Context, x apiGrant) error {
+	id, ok := x.GuardData["id"]
+	if ok {
+		if strID, ok := id.(string); !ok || !a.accessTokens.Exists(ctx, strID) {
+			return errMissingTokenID
+		}
+	} else {
+		return errMissingTokenID
+	}
+
+	guardData, err := json.Marshal(x.GuardData)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	g := authz.Grant{
+		GuardType: x.GuardType,
+		GuardData: guardData,
+		Policy:    x.Policy,
+		CreatedAt: time.Now().UTC().Format(time.RFC3339),
+	}
+
+	data, err := a.raftDB.Get(ctx, grantPrefix+x.Policy)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	if data == nil {
+		// if there aren't any grants associated with this policy, go ahead
+		// and chuck this into raftdb
+		gList := &authz.GrantList{
+			Grants: []*authz.Grant{&g},
+		}
+		val, err := proto.Marshal(gList)
+		if err != nil {
+			return errors.Wrap(err)
+		}
+		err = a.raftDB.Insert(ctx, grantPrefix+x.Policy, val)
+		if err != nil {
+			return errors.Wrap(err)
+		}
+		return nil
+	}
+
+	grantList := new(authz.GrantList)
+	err = proto.Unmarshal(data, grantList)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	grants := grantList.GetGrants()
+	for _, existing := range grants {
+		if existing.GuardType == x.GuardType && bytes.Equal(existing.GuardData, guardData) {
+			// this grant already exists, return for idempotency
+			return nil
+		}
+	}
+
+	// create new grant and it append to the list of grants associated with this policy
+	grants = append(grants, &g)
+	gList := &authz.GrantList{Grants: grants}
+	val, err := proto.Marshal(gList)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	// TODO(tessr): Make this safe for concurrent updates. Will likely require a
+	// conditional write operation for raftDB
+	err = a.raftDB.Set(ctx, grantPrefix+x.Policy, val)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (a *API) listGrants(ctx context.Context) (map[string]interface{}, error) {
+	var grants []apiGrant
+	for _, p := range policies {
+		// perhaps could denormalize the data in storage to speed this up,
+		// but for now assume a small number of grants
+		data, err := a.raftDB.Get(ctx, grantPrefix+p)
+		if err != nil {
+			return nil, errors.Wrap(err)
+		}
+		if data == nil {
+			continue
+		}
+
+		grantList := new(authz.GrantList)
+		err = proto.Unmarshal(data, grantList)
+		if err != nil {
+			return nil, errors.Wrap(err)
+		}
+		for _, g := range grantList.GetGrants() {
```

Comment:
Nit: instead of GetGrants it would be fine (and IMO slightly preferable) to use the field itself `grantList.Grants` unless we need the extra nil check on the receiver (which we don't here).